### PR TITLE
Stupid POC I was doing on your app: Make aggregates modular.

### PR DESCRIPTION
**Pull request by briantopping**
Hey Heiko. Not sure if this is at all interesting since you are using this app for training and I don't know the goals of your course. 

It seems if you are going toward DDD, being able to encapsulate domains as subprojects would be valuable on larger projects. I didn't break the `Flow` out into a separate project, but it could easily be done with the `FlowFacade` as an `Aggregate` that can be loaded as a decoupled module. This would allow the root app and the mechanics around clustering to be more easily considered a separate Bounded Context from the language used with the `Flow`. 

Feel free to reject this, I just thought it was better to create it than to have you tell me some year when we finally meet "geez, that would have been pretty cool!" At the same time, if you want me to improve the PR with a broken out Flow subproject, please do let me know!

### CODE REVIEW

File: src/main/scala/de/heikoseeberger/reactiveflows/ReactiveFlowsApp.scala [hunk 1]
```diff
@@ -16,25 +16,27 @@
...
 import akka.event.Logging
+import scala.collection.JavaConversions._
 
-object ReactiveFlowsApp {
+object ReactiveFlowsApp extends App {
 
   private val jvmArg = """-D(\S+)=(\S+)""".r
 
-  def main(args: Array[String]): Unit = {
+  override def main(args: Array[String]): Unit = {
```
Comment (hseeberger) [inline src/main/scala/de/heikoseeberger/reactiveflows/ReactiveFlowsApp.scala]: Overriding `App.main` is not a good idea, it defeats its purpose.

**Comment by hseeberger** [comment]
Hey Brian,

I don't understand DDD well enough to understand your changes. Maybe you could explain that a little bit?

Heiko

File: src/main/scala/de/heikoseeberger/reactiveflows/ReactiveFlowsApp.scala (hunk 1 above)
Comment (briantopping) [inline src/main/scala/de/heikoseeberger/reactiveflows/ReactiveFlowsApp.scala]: Ah that change wasn't supposed to be in there, will remove.

**Comment by briantopping** [comment]
Sure, the most core ideas (to me) are to get what amounts to a localized vocabulary for talking about a problem domain isolated to contained areas of the code. The "Ubiquitous Language" is one that can be shared with domain experts and the "Bounded Context" is the scope of what might be considered as an integratabtle unit. 

Separately, akka-persistence has moved toward a more complete embrace of CQRS/ES ([ref](https://groups.google.com/forum/#!msg/akka-user/MNDc9cVG1To/GiAlwC3n3aYJ)) and CQRS often indicates DDD ([ref](http://martinfowler.com/bliki/CQRS.html)). Use of a design patterns such as those in DDD are not detrimental to the understanding of the uninitiated, but to those that use them, convey much more meaning, as is always true with patterns. 

In the case of Reactive Flows, my belief is the `Facade` pattern is actually a better match for the DDD `Aggregate`, which clusters domain entities and value objects within a transactional consistency boundary. In turn, I believe these form an optimal `Module` as well. Vaughn Vernon would call a module a "named container for domain object classes that are highly cohesive with one another”. Aggregates form good candidates for modules because maintenance of transactional consistency over time benefits from lack of coupling. This is reinforced by the requirement that an `Actor` does not expose it's inner structures except through messages. 

Aggregates are also an ideal form of module as aggregates can form dependencies on each other through the value objects they exchange. The type composition of these value objects needs to be tracked in a fluent manner -- semantic versions of modules managed by the build system. 

Once we have included a module on the classpath, we should like to discover it's existence and instantiate it in the lightest weight manner as possible. The Java ServiceLoader API provides such an interface without the overhead of a full DI implementation, but does so in a manner that a full DI implementation can be substituted if it's indicated. As a fully realized system will have many `Aggregate`s that need to be instantiated as a part of the cluster initialization, such DI is really the cleanest way to avoid changes at the application level (for it's part, a DDD context that should not require modification of the `ReactiveFlowsApp` every time a new `Aggregate` is introduced).

While DI doesn't seem necessary in today's functional environment, I believe as time goes on, DI will re-emerge as functional allows developers to express exponentially more complex ideas and packaging/deployment requirements push back. I don't think this kind of DI is a bad pattern to introduce if it can be done in a lightweight manner such as this.

Bringing this all together, I believe the current form of the patch is the least-weight form of demonstrating how your `Facade` is actually an `Aggregate` and creating the ability for `Aggregate`s to be extracted as a separately managed projects that is easily instantiated in arbitrary number by the top level application. The evolution of this patch is to christen the `FlowAggregate` and make it into a separate SBT subproject, but doing so in the first step would have obscured the mechanics of loading an aggregate in a modular manner.

Apologies for the length, but we haven't worked together before so I wanted to give you a full brain dump. Cheers!

**Comment by hseeberger** [comment]
Brian,

No need for apologies – on the contrary, I totally appreciate your enthusiasm!

Actually I think I like DDD a lot – even though I don’t know it well enough. But I haven’t ever used the ServiceLoader API and I honestly have some doubts … Would it maybe make more sense to rename some classes, e.g. FlowFacade to FlowAggregate (I don’t know if that’s a good practice)? Or some other changes that introduce DDD more at the conceptual level that on a very JVM specific?

Heiko

--

Heiko Seeberger
Home: heikoseeberger.de <http://heikoseeberger.de/>
Twitter: @hseeberger <https://twitter.com/hseeberger>
Public key: keybase.io/hseeberger <https://keybase.io/hseeberger>

**Comment by briantopping** [comment]
Yes, I've been traditionally a little skeptical of this API as well, but really all it's doing is the classloader theatrics and reflection across the entire classpath and using the files in META-INF/services to map from the name of the interface (the name of the file) to one or more implementations (the lines of the file, one FQCN per line). It could easily be done in handwritten code and maybe Scala standard library should have something like this. It's a very convenient API in lightweight situations like this.

If all one ever needs is a monolithic application, none of this adds any value and not worth it. As the patch is currently, the power of it isn't evident. I left out the subproject extraction for a later iteration. I strongly believe the need to modify the main class every time there's a new Aggregate is an anti-pattern, this solves it as quietly as possible.

Not trying to pitch you on this at all, just offering some perspective. I may go back to OSGi for items like the aggregates, but for initial development, wanted to keep things lightweight. Your app is working where mine is not, so I ended up patching yours first to make sure the pattern is operational.

**Comment by briantopping** [comment]
One of the challenges I found in the last few hours of working on this is the actor supervision hierarchy. I believe it would be more useful if every module had it's own hierarchy so `Props` don't need to be propagated to higher levels in order to manage the supervision properly. I just don't know the motivations you had for the current supervision, so it's a bit of a tough call.

**Comment by briantopping** [comment]
Hi Heiko, thought I would check in with you on this. I went a bit off the reservation and generated https://github.com/briantopping/reactive-flows/tree/bt-module. Hopefully it doesn't require you to swallow your lunch a second time, this is the first time I've had to refactor someone else's Scala and I'm a little out of my element with it. 

When it's operational, the aggregate name ("flow-facade") will be a part of the path to get to routes in the aggregate. There seemed to be a lot of code bleeding between HttpService and FlowFacade, I think this separation would help keep new modules very clean. I'm quite sure there's a lot of new broken stuff in there, stuff moved to the wrong places and values from the `Settings` that were very poorly ported.

WDYT of the general pattern? I think a lot of it is visible without even opening any files. Grateful for any feedback you can provide!!

**Comment by hseeberger** [comment]
Hi Brian,

I still have a hard time getting my head around the core idea behind this refactoring. You mentioned code bleeding btw HttpService and FlowFacade. Can you give examples? Then, what exactly have you done to minimise this coupling – as well conceptually (don't hesitate using DDD terms, I want to learn) as code-wise.

Thanks
Heiko

**Comment by briantopping** [comment]
Hi good morning Heiko,

I'll put together a stripped down example of it. I think it's going to be too difficult to explain any better than I have and there's too many moving parts to work around in what you've done. So I kind of need to rebuild it from scratch, I just don't have the Scala-foo to refactor what you've done yet, even though it's not that big. (Which has been an educational experience for me because I see now how hard it could be in an enterprise).

Sorry for eating up your time on this, I thought this might get the idea across but it wasn't enough yet.

**Comment by hseeberger** [comment]
Don't worry, I think aiming for decoupling via DDD (or Mircoservices or ...) is a great idea.

**Comment by briantopping** [comment]
Thanks for the confidence :) Something will come of it.

I'll reopen as another ticket when I have something, cheers!

**Pull request closed without merging**